apps/spark/src/test/java/com/linkedin/openhouse/jobs/spark/OperationsTest.java

```diff
@@ -286,14 +322,81 @@ public void testSnapshotsExpirationJavaAPI() throws Exception {
+
+      ops.expireSnapshots(table, maxAge, timeGranularity, 0);
+      // No snapshots should be cleaned up as they are all within the max age
+      checkSnapshots(table, snapshotIds);
+    }
+    // restart the app to reload catalog cache
+    try (Operations ops = Operations.withCatalog(getSparkSession(), meter)) {
+      // verify that new apps see snapshots correctly
+      checkSnapshots(ops, tableName, snapshotIds);
+    }
+  }
+
+  @Test
+  public void testSnapshotsExpirationVersions() throws Exception {
+    final String tableName = "db.test_es_versions_java";
+    final int numInserts = 3;
+    final int versionsToKeep = 2;
+    List<Long> snapshotIds;
+    try (Operations ops = Operations.withCatalog(getSparkSession(), meter)) {
+      prepareTable(ops, tableName);
+      populateTable(ops, tableName, numInserts);
+      snapshotIds = getSnapshotIds(ops, tableName);
+      Assertions.assertEquals(
+          numInserts,
+          snapshotIds.size(),
+          String.format("There must be %d snapshot(s) after inserts", numInserts));
+      Table table = ops.getTable(tableName);
+      log.info("Loaded table {}, location {}", table.name(), table.location());
+
+      ops.expireSnapshots(table, 0, "", versionsToKeep);
+      // verify that table object snapshots are updated
+      checkSnapshots(
+          table, snapshotIds.subList(snapshotIds.size() - versionsToKeep, snapshotIds.size()));
+    }
+    // restart the app to reload catalog cache
+    try (Operations ops = Operations.withCatalog(getSparkSession(), meter)) {
+      // verify that new apps see snapshots correctly
+      checkSnapshots(
+          ops,
+          tableName,
+          snapshotIds.subList(snapshotIds.size() - versionsToKeep, snapshotIds.size()));
+    }
+  }
+
+  @Test
+  public void testSnapshotsExpirationBothAgeAndVersions() throws Exception {
+    final String tableName = "db.test_es_age_and_versions_java";
+    final int numInserts = 3;
+    final int maxAge = 3;
+    final String timeGranularity = "DAYS";
+    final int versionsToKeep = 1;
+    List<Long> snapshotIds;
+    try (Operations ops = Operations.withCatalog(getSparkSession(), meter)) {
+      prepareTable(ops, tableName);
+      populateTable(ops, tableName, numInserts);
+      snapshotIds = getSnapshotIds(ops, tableName);
+      Assertions.assertEquals(
+          numInserts,
+          snapshotIds.size(),
+          String.format("There must be %d snapshot(s) after inserts", numInserts));
+      Table table = ops.getTable(tableName);
+      log.info("Loaded table {}, location {}", table.name(), table.location());
+
+      ops.expireSnapshots(table, maxAge, timeGranularity, versionsToKeep);
       // verify that table object snapshots are updated
-      checkSnapshots(table, snapshotIds.subList(2, snapshotIds.size()));
+      checkSnapshots(
+          table, snapshotIds.subList(snapshotIds.size() - versionsToKeep, snapshotIds.size()));
```

Comment:
how does it test the semantics of favoring maxAge when both spec coexists? I assume we need to create N + M  snapshots in total, with N > versionToKeep and all snapshots among N are oldThanMaxAge, and after calling expiration job we will have all N expired.

Reply:
Thanks, I added some tests around this and found a bug where `retainLast` would throw an exception if the `versionToKeep > N`